### PR TITLE
do @ variable substitution when parsing cmake format

### PR DESCRIPTION
**Pull request by Jan200101**
the upstream behavior of configure_file in cmake parses both @VAR@ and ${VAR} and only supports disabling the bracket variables through the `@ONLY` argument, meson needs to follow this behavior for compatibility

this is an ugly solution but I think its best considering the alternatives:
- calling `do_conf_str` twice with both formats
    - will break if the first iteration places substitutes that match the format of the second iteration
- adapt the existing bracket format regex to support @
    - the existing regex is unholy and broken in more ways than I am willing to list, it would be better to completely gut this part of the code and replace it with regular python code
 - rewrite the code to not use regex
     - this would be quite complex and I am not familiar enough with mesondefine to be able to find any regressions. The lack of specification on how its suppose to work does not help either

### CODE REVIEW

File: mesonbuild/utils/universal.py [hunk 1]
```diff
@@ -1280,11 +1283,15 @@ def get_variable_regex(variable_format: Literal['meson', 'cmake', 'cmake@'] = 'm
         ''', re.VERBOSE)
     else:
         regex = re.compile(r'''
-            (?:\\\\)+(?=\\?\$)  # Match multiple backslashes followed by a dollar sign
+            (?:\\\\)+(?=\\?(\$|@))  # Match multiple backslashes followed by a dollar sign
```
Comment (dcbaker) [inline mesonbuild/utils/universal.py:1286]: looks like the comment should be updated

**Review by dcbaker — CHANGES_REQUESTED** [review]
Thanks for looking at this, I had noticed this but never had time to fix it.

It would be really good to get test coverage fro the `cmake@` format, in `test cases/common/14 configure file`

File: mesonbuild/utils/universal.py (hunk 1 above)
Comment (Jan200101) [reply mesonbuild/utils/universal.py:1286]: Actually I need to revert this, it was part of the attempt at updating the regex which ultimately proved to be unreasonable

Comment (Jan200101) [reply mesonbuild/utils/universal.py:1286]: reinspecting this with a less tired mind, the change is actually right and I do need to update the comment

Comment (Jan200101) [reply mesonbuild/utils/universal.py:1286]: resolved by 5c758763bf0a8bf63099701dfa96525a7ae7f048

**Comment by Jan200101** [comment]
rebasing to un-screw CI

**Review by dcbaker — APPROVED** [review]

**Comment by eli-schwartz** [comment]
> rebasing to un-screw CI
> EDIT: it ended up becoming worse, somehow, I don't think this is caused by this PR

```
mkenums/main6.c:8:42: error: ‘MESON_TYPE_THE_XENUM’ undeclared (first use in this function)
```

This is not a known or preexisting failure. I would venture to say something about this PR is in conflict with behavior that that test case was relying on, but I haven't had a chance to review the implementation.

**Comment by eli-schwartz** [comment]
We have a file, enums.h.in:
https://github.com/mesonbuild/meson/blob/8156e12ce0de62aa47472e529489f7412628e6e5/test%20cases/frameworks/7%20gnome/mkenums/enums.h.in#L7-L21

We convert it to enums6.h.in and then use it as a template for `glib-mkenums`:
https://github.com/mesonbuild/meson/blob/8156e12ce0de62aa47472e529489f7412628e6e5/test%20cases/frameworks/7%20gnome/mkenums/meson.build#L143-L155

The replacement tokens must be parsed by glib-mkenums, not configure_file. We use configure_file in order to copy a file without modifications, and that's what this PR causes to stop behaving the way it used to behave. If you configure the test case before and after this PR:

File created by glib-mkenums:
```diff
diff --git a/builddir-old/mkenums/enums6.h b/builddir-new/mkenums/enums6.h
index 313d846ff..950175623 100644
--- a/builddir-old/mkenums/enums6.h
+++ b/builddir-new/mkenums/enums6.h
@@ -8,11 +8,11 @@
 
 G_BEGIN_DECLS
 
-/* enumerations from "meson-sample.h" */
-GType meson_the_xenum_get_type(void) G_GNUC_CONST;
-#define MESON_TYPE_THE_XENUM (meson_the_xenum_get_type())
-GType meson_the_flags_enum_get_type(void) G_GNUC_CONST;
-#define MESON_TYPE_THE_FLAGS_ENUM (meson_the_flags_enum_get_type())
+/* enumerations from "" */
+GType _get_type(void) G_GNUC_CONST;
+#define _TYPE_ (_get_type())
+GType _get_type(void) G_GNUC_CONST;
+#define _TYPE_ (_get_type())
 
 G_END_DECLS
 
```

File created by configure_file:
```diff
diff --git a/builddir-old/mkenums/enums6.h.in b/builddir-new/mkenums/enums6.h.in
index 479867fd1..7f6a50107 100644
--- a/builddir-old/mkenums/enums6.h.in
+++ b/builddir-new/mkenums/enums6.h.in
@@ -9,11 +9,11 @@ G_BEGIN_DECLS
 
 /*** BEGIN file-production ***/
 
-/* enumerations from "@basename@" */
+/* enumerations from "" */
 /*** END file-production ***/
 /*** BEGIN value-header ***/
-GType @enum_name@_get_type(void) G_GNUC_CONST;
-#define @ENUMPREFIX@_TYPE_@ENUMSHORT@ (@enum_name@_get_type())
+GType _get_type(void) G_GNUC_CONST;
+#define _TYPE_ (_get_type())
 /*** END value-header ***/
 
 /*** BEGIN file-tail ***/
```

**Comment by eli-schwartz** [comment]
I think what this means is that the existing testcase code was relying on `format: 'cmake'` not actually implementing cmake replacements, and obviously implementing that missing feature meant that we then mangled the file by, well, copying it.

Or something.

Those `@VARNAME@` tokens weren't part of the empty configuration_data used, obviously, so they ended up being deleted.

**Review by dcbaker — APPROVED** [review]

**Pull request merged**